### PR TITLE
Add conformance testing DSL

### PR DESCRIPTION
**Pull request by nirosys**
*Issue #, if available:* n/a

*Description of changes:*
This PR adds an implementation of the conformance DSL, along with a runner for testing conformance tests during CI/CD, and a separate runner that can be used to target files more ad-hoc. The ion-tests submodule has also been updated in order to pull in the conformance tests, which also required some updates to the `test_resource` paths for 1.0.

Most of the code is modeled directly after the grammar. Documents contain fragments and a continuation. Expectations, which are a subset of continuations, act as terminal nodes and trigger the evaluation of the test. Context tracks the fragments, version, and encoding, for the test's evaluation path.

## CI/CD Runner
```
$ cargo test -F experimental --test conformance_tests
   Compiling ion-rs v1.0.0-rc.7 (/Users/glitch/Code/ion-rust)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 3.65s
     Running tests/conformance_tests.rs (target/debug/deps/conformance_tests-6a032b27940772cd)

running 8 tests
test ion_tests::conformance_ion_tests_conformance_null_ion ... ok
test implementation::test_encoding ... ok
test ion_tests::conformance_ion_tests_conformance_core_string_symbol_ion ... ok
test implementation::test_timestamps ... ok
test ion_tests::conformance_ion_tests_conformance_core_empty_document_ion ... ok
test implementation::test_simple_docs ... ok
test ion_tests::conformance_ion_tests_conformance_core_typed_null_ion ... ok
test ion_tests::conformance_ion_tests_conformance_core_toplevel_produces_ion ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s

$
```

## CLI Runner
```
$ cat test.ion
(ion_1_1 "Test Bytes" (text "{{aGVsbG8=}}") (denotes (Blob "68 65 6c 6c 6f")))
(ion_1_1 "Test Bytes (Fail)" (text "{{aGVsbG8=}}") (denotes (Blob "68 65 6c 6c 5f")))

(ion_1_1 "Test Clob" (text "{{ \"hello\" }}") (denotes (Clob "68 65 6c 6c 6f")))
(ion_1_1 "Test Clob (Fail)" (text "{{ \"hello\" }}") (denotes (Clob "68 65 6c 6c 5f")))

$ cargo test -F experimental --test conformance -- ./test.ion
   Compiling ion-rs v1.0.0-rc.7 (/Users/glitch/Code/ion-rust)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 1.92s
     Running tests/conformance.rs (target/debug/deps/conformance-62318380ee1917b3)
Testing 1 conformance collections.


Running tests: ./test.ion ========================
   Test Bytes         ...  [OK]
   Test Bytes (Fail)  ...  [FAILED]
   Test Clob          ...  [OK]
   Test Clob (Fail)   ...  [FAILED]
-------------------------
File: ./test.ion
Test: Test Bytes (Fail)
Error: ConformanceError(ConformanceErrorImpl { file: "", test_name: "", kind: MismatchedDenotes })
-------------------------
File: ./test.ion
Test: Test Clob (Fail)
Error: ConformanceError(ConformanceErrorImpl { file: "", test_name: "", kind: MismatchedDenotes })
thread 'main' panicked at tests/conformance.rs:51:9:
Conformance test(s) failed
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `--test conformance`
$
```

## Notes
* The MacTab fragment is implemented as the spec describes, expanding into a toplevel fragment with named module for defining the macro table. Adding a new named module is not supported in ion-rust yet. Macros can still be added using the encoding fragment.
* Absent symbols are currently not supported. 

I'll follow up after this PR to make errors reporting better. I also have some plans for refactoring and simplifying some of the implementation, post merge.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lazy/binary/raw/v1_1/value.rs [hunk 1]
```diff
@@ -194,7 +194,13 @@ impl<'top> LazyRawValue<'top, BinaryEncoding_1_1> for &'top LazyRawBinaryValue_1
...
         }
         if self.is_null() {
-            return Ok(ValueRef::Null(self.ion_type()));
+            let ion_type = if self.encoded_value.header.ion_type_code == OpcodeType::TypedNull {
+                let body = self.value_body();
+                ION_1_1_TYPED_NULL_TYPES[body[0] as usize]
+            } else {
+                IonType::Null
+            };
+            return Ok(ValueRef::Null(ion_type));
```
Comment (zslayton) [inline src/lazy/binary/raw/v1_1/value.rs:203]: Why was this needed? `ion_type()` should return the correctly interpreted type from the opcode and optional typed null byte. (`return Ok(ValueRef::Null(self.ion_type()))` also appears below on line 227).

File: tests/conformance_dsl/context.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            let empty: Vec<Element> = vec!();
+            return Ok(empty.into());
+        }
+
+        let version = match self.version() {
+            IonVersion::Unspecified => IonVersion::V1_0,
+            v => v,
+        };
+
+        // Ok(Reader::new(AnyEncoding, data_slice)?.read_all_elements()?)
```
Comment (zslayton) [inline tests/conformance_dsl/context.rs:108]: Old comment?
```suggestion
        // Ok(Reader::new(AnyEncoding, data_slice)?.read_all_elements()?)
```

File: tests/conformance_dsl/model.rs [hunk 3]
```diff
@@ -0,0 +1,411 @@
+use ion_rs::{Decimal, Element, IonType, Sequence, Timestamp};
+use ion_rs::decimal::coefficient::Coefficient;
+use super::{Clause, ClauseType, ConformanceErrorKind, InnerResult, parse_text_exp, parse_bytes_exp};
+
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq)]
+pub(crate) enum SymTok {
```
Comment (zslayton) [inline tests/conformance_dsl/model.rs:8]: I suggest spelling out the words:
```suggestion
pub(crate) enum SymbolToken {
```

File: tests/conformance_tests.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+mod ion_tests {
+    use super::*;
+
+    #[test_resources("ion-tests/conformance/null.ion")]
+    #[test_resources("ion-tests/conformance/core/typed_null.ion")]
+    #[test_resources("ion-tests/conformance/core/string_symbol.ion")]
+    #[test_resources("ion-tests/conformance/core/empty_document.ion")]
+    #[test_resources("ion-tests/conformance/core/toplevel_produces.ion")]
+    // #[test_resources("ion-tests/conformance/ivm.ion")]
```
Comment (zslayton) [inline tests/conformance_tests.rs:86]:
```suggestion
    // #[test_resources("ion-tests/conformance/ivm.ion")]
```
?

**Review by zslayton — COMMENTED** [review]
This looks good! Please do a documentation pass. It would be helpful to connect implementation sections to the part of the DSL spec they implement. Having an explanation of the motivation for top-level types like `ModelValue` would also be nice.

File: src/lazy/binary/raw/v1_1/value.rs (hunk 1 above)
Comment (nirosys) [reply src/lazy/binary/raw/v1_1/value.rs:203]: When parsing the Opcode we don't have the byte containing the true ion type ([here](https://github.com/amazon-ion/ion-rust/blob/main/src/lazy/binary/raw/v1_1/type_descriptor.rs#L79)), so the actual ion type was originally deferred until the value is read ([here](https://github.com/amazon-ion/ion-rust/blob/main/src/lazy/binary/raw/v1_1/value.rs#L147)). Right now `ion_type()` just returns the type from the opcode, but you're right, it's a better place to be handling this logic.

File: tests/conformance_tests.rs (hunk 4 above)
Comment (nirosys) [reply tests/conformance_tests.rs:86]: Yea, I had commented this out because it currently doesn't pass due to the symbol syntax that's needed. Removing it for now.

File: src/lazy/binary/raw/v1_1/value.rs [hunk 5]
```diff
@@ -119,7 +119,13 @@ impl<'top> HasRange for &'top LazyRawBinaryValue_1_1<'top> {
...
 impl<'top> LazyRawValue<'top, BinaryEncoding_1_1> for &'top LazyRawBinaryValue_1_1<'top> {
     fn ion_type(&self) -> IonType {
-        self.encoded_value.ion_type()
+        // Handle retrieving the type for a typed null.
+        if self.encoded_value.header.type_code() == OpcodeType::TypedNull {
+            let body = self.value_body();
+            ION_1_1_TYPED_NULL_TYPES[body[0] as usize]
+        } else {
+            self.encoded_value.ion_type()
+        }
```
Comment (zslayton) [inline src/lazy/binary/raw/v1_1/value.rs:128]: Could you put this functionality in a `self.read_null()` method instead of having it happen in `ion_type()`?

There are three other raw value types besides `LazyRawBinaryValue_1_1`:
* `LazyRawBinaryValue_1_0`
* `LazyRawTextValue_1_1`
* `LazyRawTextValue_1_0`

Each of them has an `ion_type()` method and I'm pretty sure at least some of them return `IonType::Null` if the value is a null value, even if it's a typed null. The behavior you've implemented is probably more correct, but if we make this change here we'd need to make it in the other three places too.

**Review by zslayton — APPROVED** [review]

File: src/lazy/binary/raw/v1_1/value.rs (hunk 5 above)
Comment (zslayton) [reply src/lazy/binary/raw/v1_1/value.rs:128]: Added #828 to track this.

Comment (nirosys) [reply src/lazy/binary/raw/v1_1/value.rs:128]: ack, ty, I will follow up with that today.
